streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java

```diff
@@ -679,58 +675,75 @@ private void cleanupTask(final Task task) {
                         final Map<TopicPartition, OffsetAndMetadata> committableOffsets = task.prepareCommit();
+                        tasksToCommit.add(task);
                         if (task.isActive()) {
                             consumedOffsetsAndMetadataPerTask.put(task.id(), committableOffsets);
                         }
                     }
-                    tasksToClose.add(task);
+                    tasksToCloseClean.add(task);
                 } catch (final TaskMigratedException e) {
                     // just ignore the exception as it doesn't matter during shutdown
-                    closeTaskDirty(task);
+                    tasksToCloseDirty.add(task);
                 } catch (final RuntimeException e) {
                     firstException.compareAndSet(null, e);
-                    closeTaskDirty(task);
+                    tasksToCloseDirty.add(task);
                 }
-            } else {
-                closeTaskDirty(task);
             }
-        }
 
-        try {
-            if (clean) {
-                commitOffsetsOrTransaction(consumedOffsetsAndMetadataPerTask);
-                for (final Task task : tasksToCommit) {
-                    try {
-                        task.postCommit();
-                    } catch (final RuntimeException e) {
-                        log.error("Exception caught while post-committing task " + task.id(), e);
-                        firstException.compareAndSet(null, e);
-                    }
-                }
+            // If any active tasks have to be clsoed dirty and can't be committed, none of them can be
+            if (!filterActive(tasksToCloseDirty).isEmpty()) {
+                tasksToCloseClean.removeAll(filterActive(tasksToCommit));
+                tasksToCommit.removeAll(filterActive(tasksToCommit));
+                tasksToCloseDirty.addAll(activeTaskIterable());
+                consumedOffsetsAndMetadataPerTask.clear();
             }
-        } catch (final RuntimeException e) {
-            log.error("Exception caught while committing tasks during shutdown", e);
-            firstException.compareAndSet(null, e);
-        }
 
-        for (final Task task : tasksToClose) {
             try {
-                completeTaskCloseClean(task);
+                commitOffsetsOrTransaction(consumedOffsetsAndMetadataPerTask);
             } catch (final RuntimeException e) {
+                log.error("Exception caught while committing tasks during shutdown", e);
                 firstException.compareAndSet(null, e);
-                closeTaskDirty(task);
+
+                // If the commit fails, everyone who participated in it must be closed dirty
+                tasksToCloseDirty.addAll(filterActive(tasksToCommit));
+                tasksToCloseClean.removeAll(filterActive(tasksToCommit));
+                tasksToCommit.clear();
+            }
+
+            for (final Task task : tasksToCommit) {
```

Comment:
Regarding `If any active tasks have to be closed dirty and can't be committed, none of them can be`, is that only for eos-beta?

Reply:
I guess, technically it's only required for eos-beta, but we to avoid too many different cases, we might just want to do the same thing for all cases?

Reply:
Yeah I agree with Matthias, with eos-alpha we don't currently differentiate which tasks may have succeeded the commit and. which may have failed, so this would require a larger refactoring. My take is that it doesn't really matter much, since practically speaking it's unlikely that one task would successfully commit while other tasks would fail
(ack on moving into the same try-catch)